torch/distributed/fsdp/_state_dict_utils.py

```diff
@@ -31,49 +38,218 @@
 from .flat_param import FlatParamHandle
 
 
-def _full_post_state_dict_hook(
+def _enter_full_param_ctx(
+    module,
+    recurse: bool = False,
+    writeback: bool = False,
+    rank0_only: bool = False,
+    offload_to_cpu: bool = False,
+    with_grads: bool = False,
+) -> None:
+    """
+    state_dict hooks cannot use the pure context call as the checkpoint flow
+    requires to enter the context in the pre-hook but leave the context in the
+    post-hook. This API enters the context of ``summon_full_params``.
+    """
+    assert module._full_param_ctx is None, (
+        "Entering the ``summon_full_params`` context but module._full_param_ctx "
+        "is not None."
+    )
+    assert module.training_state != TrainingState.SUMMON_FULL_PARAMS, (
+        "Entering the summon_full_params context but the state is already "
+        "SUMMON_FULL_PARAMS."
+    )
+    module._full_param_ctx = module._summon_full_params(
+        recurse=recurse,
+        writeback=writeback,
+        rank0_only=rank0_only,
+        offload_to_cpu=offload_to_cpu,
+        with_grads=with_grads,
+    )
+    module._full_param_ctx.__enter__()
+
+
+def _exit_full_param_ctx(module) -> None:
+    """A helper function to exit ``summon_full_params`` context."""
+    module._assert_state([TrainingState.SUMMON_FULL_PARAMS])
+    assert module._full_param_ctx is not None
+    module._full_param_ctx.__exit__(None, None, None)
+    module._full_param_ctx = None
+
+
+def _common_pre_state_dict_hook(
+    module,
+    state_dict: Dict[str, Any],
+    prefix: str,
+) -> None:
+    """Performs the pre-state_dict tasks shared by all state_dict types."""
+    if torch.cuda.is_available():
+        torch.cuda.synchronize()
+    _lazy_init(module, module)
+    # TODO: change to this call after pre_state_dict_hook is in `nn.Module`.
+    # if module.is_root:
+    #    _clear_grads_if_needed(module._fsdp_handles(module))
+    if module._has_params:
```

Comment:
Sorry, let me clarify the `_clear_grads_if_needed()` call.

The idea is that we want to free the gradient memory as early as possible, so, given that this `_clear_grads_if_needed()` incurs minimal overhead (looping over gradients and checking if all are `None`), I inserted this call at every FSDP callsite.

In that way, we should prefer to keep it as is since the root can help free all possible gradient memory, which may be earlier than deferring to the submodule.

Reply:
Right now, the pre_hook is not an actual pre_state_dict_hook. So it will be called from the bottom-up. So the gradient won't be freed in time for a leaf module before the leaf module's post_state_dict_hook is called. Once the pre_state_dict_hook is actually implemented in `nn.Module`, I'll change it back.